resource/resourceadapters/opener.go

```diff
@@ -108,6 +115,49 @@ func (ro *ResourceOpener) OpenResource(name string) (o resource.Opened, err erro
 	return opened, nil
 }
 
+// resourceState adapts between resource state and charmstore.EntityCache.
+type resourceState struct {
+	st      Resources
+	userID  names.Tag
+	unit    resource.Unit
+	appName string
+}
+
+// GetResource implements charmstore.EntityCache.
+func (s *resourceState) GetResource(name string) (resource.Resource, error) {
+	return s.st.GetResource(s.appName, name)
+}
+
+// SetResource implements charmstore.EntityCache.
+func (s *resourceState) SetResource(chRes charmresource.Resource, reader io.Reader) (resource.Resource, error) {
+	return s.st.SetResource(s.appName, s.userID.Id(), chRes, reader)
+}
+
+// OpenResource implements charmstore.EntityCache.
+func (s *resourceState) OpenResource(name string) (resource.Resource, io.ReadCloser, error) {
+	if s.unit == nil {
+		return resource.Resource{}, nil, errors.NotImplementedf("")
+	}
+	return s.st.OpenResourceForUniter(s.unit, name)
+}
+
+type nopOpener struct{}
+
+func newNopOpener() *nopOpener {
+	return &nopOpener{}
+}
+
+// NewClient opens a new charm store client.
+func (o *nopOpener) NewClient() (*ResourceRetryClient, error) {
+	return newRetryClient(nopClient{}), nil
+}
+
+type nopClient struct{}
+
+func (nopClient) GetResource(req repositories.ResourceRequest) (charmstore.ResourceData, error) {
```

Comment:
Comment here to set the scene of it's usage. I.e. that _not found_ means we can only get it from state.